### ISSUE TITLE
Define ODO user/scope && few questions

### ISSUE THREAD

**Issue by surajnarwade** [openshift/odo#722]
I have few questions:

* who is odo user ? 
* does he know openshift in advance ?
* does he care about yamls in openshift ?
----------------------------------------------------------------
* who is superuser ?
* will superuser use odo ?
------------------------------------------------------------------
* do we need to consider that user will use both oc and odo ?
* for greater user experience, do we need to draw some line between what we will provide and what not ? (just had a fear that, eventually odo should not become oc again )
-------------------------------------------------------------------
* I was thinking this way:
    * if you are ops, use `oc`
    * if you are noob/dev, use `odo`
Am I correct here ?

**Comment by surajnarwade** [comment]
cc @kadel @cmoulliard @sspeiche

**Comment by anmolbabu** [comment]
Some of the questions also from the discussion in :

https://github.com/redhat-developer/odo/pull/626#discussion_r215315130

may be relevant to this

**Comment by jorgemoralespou** [comment]
Here are `my` answers, which can differ from @sspeiche

>    who is odo user ?

A developer that is not (nor want/need to be) familiar with OpenShift and kubernetes but needs to create applications using this platform

>    does he know openshift in advance ?

Don't need to, and the ideal user is the one not really experienced with OpenShift

>    does he care about yamls in openshift ?

For sure not

>    who is superuser ?

What is a superuser?

>    will superuser use odo ?

Since I don't know what you mean by superuser, can't answer

>    do we need to consider that user will use both oc and odo ?

Ideally, our users would only use `odo, but there will be many users, those familiar with OpenShift that for convenience might use both

>    for greater user experience, do we need to draw some line between what we will provide and what not ? (just had a fear that, eventually odo should not become oc again )

Hopefully, when oc will be implemented as plugins, we should consider doing same/similar so we can use same plugins or embed/vendor their code into our own plugin. Think that there will be differences in our commands, order, options, etc... for many things, but for others they could be just drop in (or almost). This is one of the reasons I introduced the plugin mechanism the openshift guys are thinking so that we can start trying to align the sooner the better.

>    I was thinking this way:
>       if you are ops, use oc
>       if you are noob/dev, use odo
>    Am I correct here ?

Basically yes, and there will be some that will use any interchangeably. Again, we need to start R&D about the plugin mechanism in kubectl/openshift and when we will be able to pick it up and how.

**Comment by surajnarwade** [comment]
@lalatendum, can you please elaborate on superuser ?

**Comment by sspeiche** [comment]
> Here are `my` answers, which can differ from @sspeiche
> 
> > who is odo user ?
> 
> A developer that is not (nor want/need to be) familiar with OpenShift and kubernetes but needs to create applications using this platform

Agree. There are many kinds of developers. odo is focused on developers who just want to use typical concepts and flows they would expect from a "PaaS"-like platform. Abstract away some of the complexities of the platform and get out of my way to get my development done.

> 
> > does he know openshift in advance ?
> 
> Don't need to, and the ideal user is the one not really experienced with OpenShift

Another way of saying it, we should not require any knowledge of OpenShift prior to using odo, other than it is a thing, you need connection info and go.

> 
> > does he care about yamls in openshift ?
> 
> For sure not
Agree, should only care for certain advanced use cases.

> 
> > who is superuser ?
> 
> What is a superuser?
I assume you mean "cluster-admin" or someone who really knows OpenShift. This tool isn't designed for them but if they find it useful to do their job, by all means, go for it.

> 
> > will superuser use odo ?
> 
> Since I don't know what you mean by superuser, can't answer
> 
> > do we need to consider that user will use both oc and odo ?
> 
> Ideally, our users would only use `odo`, but there will be many users, those familiar with OpenShift that for convenience might use both
Another way I look at it, `odo` will hit a certain number of use cases and when it comes down to some of the advanced or not covered ones, the user may need to drop and do things like `oc` and YAML.

> 
> > for greater user experience, do we need to draw some line between what we will provide and what not ? (just had a fear that, eventually odo should not become oc again )
> 
> Hopefully, when oc will be implemented as plugins, we should consider doing same/similar so we can use same plugins or embed/vendor their code into our own plugin. Think that there will be differences in our commands, order, options, etc... for many things, but for others they could be just drop in (or almost). This is one of the reasons I introduced the plugin mechanism the openshift guys are thinking so that we can start trying to align the sooner the better.

There will be some overlap but we need to be clear and deliberate. We don't want to invest >1 time in a piece of code for Red Hat and the community to have to maintain. Based on the new plugin work in 1.12, `kubectl` is providing some libraries for plugins others to make it easier to use in their CLIs since they will need to be go binaries and may need to reimplement many of the same things. `oc` should follow that same approach and `odo` should leverage it.

There may become a day where `odo` could be a plugin to `oc` and you could do something like `oc do` or whatever, though `odo` would need to mature some but it is good to work with this in mind.

> 
> > I was thinking this way:
> > if you are ops, use oc
> > if you are noob/dev, use odo
> > Am I correct here ?
> 
> Basically yes, and there will be some that will use any interchangeably. Again, we need to start R&D about the plugin mechanism in kubectl/openshift and when we will be able to pick it up and how.

I think this is a good way at looking at it. I would say even reword slightly...
- if you are a dev, whether noob or just want to focus on dev, use odo

Good set of questions, thanks

**Comment by surajnarwade** [comment]
thanks @jorgemoralespou  and @sspeiche for answers :smile:

**Comment by cdrage** [comment]
Could we close this @surajnarwade :100:

**Comment by surajnarwade** [comment]
Closing this issue since I got my answers :tada:
